tests/Unit/ParallelAlgorithms/ApparentHorizonFinder/Criteria/Test_Residual.cpp

```diff
@@ -0,0 +1,289 @@
+  test_criterion_evaluation<Frame::Inertial>(
+      criterion, box, empty_cache, min_l_allowed_by_strahlkorper,
+      2.0 * max_residual, min_l_allowed_by_strahlkorper + 1);
+
+  // Edge case - min_resolution_l == max_resolution_l
+  const std::unique_ptr<ah::Criterion> fixed_criterion{
+      std::make_unique<Residual>(min_residual, max_residual, mid_l, mid_l)};
+  test_criterion_evaluation<Frame::Inertial>(fixed_criterion, box, empty_cache,
+                                             mid_l, 0.5 * min_residual, mid_l);
+
+  // Edge case - min_resolution_l == max_resolution_l but
+  // different from strahlkorper
+#ifdef SPECTRE_DEBUG
+  const ylm::Strahlkorper<Frame::Inertial> debug_strahlkorper{
+      make_strahlkorper<Frame::Inertial>(mid_l - 2)};
+  const FastFlow::IterInfo debug_iter_info{make_iter_info(0.5 * min_residual)};
+  CHECK_THROWS_WITH(fixed_criterion->evaluate(
+                        box, empty_cache, debug_strahlkorper, debug_iter_info),
+                    Catch::Matchers::ContainsSubstring(
+                        "If MinResolutionL == MaxResolutionL"));
+#endif
+
+  // Very large l_max
+  const size_t very_large_l{100};
+  test_criterion_evaluation<Frame::Inertial>(criterion, box, empty_cache,
+                                             very_large_l, 0.5 * min_residual,
+                                             very_large_l - 1);
+
+  // Very small residual
+  const double very_small_residual{1.0e-12};
+  test_criterion_evaluation<Frame::Inertial>(criterion, box, empty_cache, mid_l,
+                                             very_small_residual, mid_l - 1);
+
+  // Very large residual
+  const double very_large_residual{1.0};
+  test_criterion_evaluation<Frame::Inertial>(criterion, box, empty_cache, mid_l,
+                                             very_large_residual, mid_l + 1);
+
+  // Test with different frame
+  test_criterion_evaluation<Frame::Grid>(criterion, box, empty_cache, mid_l,
+                                         0.5 * min_residual, mid_l - 1);
+
+  // Test serialization
+  {
+    const Residual original{min_residual, max_residual, min_resolution_l,
+                            max_resolution_l};
+    const auto serialized{serialize_and_deserialize(original)};
+
+    // Test that the serialized object produces the same results
+    const ylm::Strahlkorper<Frame::Inertial> strahlkorper{
+        make_strahlkorper<Frame::Inertial>(mid_l)};
+    const FastFlow::IterInfo iter_info{make_iter_info(in_range_residual)};
+
+    const size_t recommended_l_original{
+        original.operator()(empty_cache, strahlkorper, iter_info)};
+    const size_t recommended_l_serialized{
+        serialized.operator()(empty_cache, strahlkorper, iter_info)};
+    CHECK(recommended_l_original == recommended_l_serialized);
+  }
+}
+
+// Test constructor validation
+SPECTRE_TEST_CASE("Unit.ApparentHorizonFinder.Criteria.Residual.Constructor",
+                  "[Unit][ParallelAlgorithms]") {
```

Comment:
Instead of having two `SPECTRE_TEST_CASE`s, put the code for each into their own free functions and just have one `SPECTRE_TEST_CASE` names `"Unit.ApparentHorizonFinder.Criteria.Residual"` that calls the two free functions.